### PR TITLE
[SPARK-30998][SQL] ClassCastException when a generator having nested inner generators

### PR DESCRIPTION
**Pull request by maropu**
### What changes were proposed in this pull request?

A query below failed in the master;

```
scala> sql("select array(array(1, 2), array(3)) ar").select(explode(explode($"ar"))).show()
20/03/01 13:51:56 ERROR Executor: Exception in task 0.0 in stage 0.0 (TID 0)/ 1]
java.lang.ClassCastException: scala.collection.mutable.ArrayOps$ofRef cannot be cast to org.apache.spark.sql.catalyst.util.ArrayData
	at org.apache.spark.sql.catalyst.expressions.ExplodeBase.eval(generators.scala:313)
	at org.apache.spark.sql.execution.GenerateExec.$anonfun$doExecute$8(GenerateExec.scala:108)
	at scala.collection.Iterator$$anon$11.nextCur(Iterator.scala:484)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:490)
	at scala.collection.Iterator$ConcatIterator.hasNext(Iterator.scala:222)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
    ...
```

This pr modified the `hasNestedGenerator` code in `ExtractGenerator` for correctly catching nested inner generators.

### Why are the changes needed?

A bug fix.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Added tests.

### CODE REVIEW

**Comment by maropu** [comment]
@WeichenXu123 @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -2164,10 +2164,19 @@ class Analyzer(
     }
 
     private def hasNestedGenerator(expr: NamedExpression): Boolean = {
+      def hasUnsupportedInnerGenerator(g: Generator): Boolean = g match {
+        case go: GeneratorOuter => // The only case that can have an inner generator
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: seems `GeneratorOuter` is just a wrapper not a real `Generator`, so it doesn't count when checking nested generators. Shall we make the comment more clear?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -2164,10 +2164,19 @@ class Analyzer(
...
+          hasUnsupportedInnerGenerator(go.child)
+        case _ =>
+          g.children.exists { _.find {
+            case _: Generator => true
+            case _ => false
+          }.isDefined }
+      }
       CleanupAliases.trimNonTopLevelAliases(expr) match {
-        case UnresolvedAlias(_: Generator, _) => false
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: so the method is named `hasNestedGenerator` but never check nested generator?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TestRelations.scala [hunk 3]
```diff
@@ -62,6 +62,9 @@ object TestRelations {
   val listRelation = LocalRelation(
     AttributeReference("list", ArrayType(IntegerType))())
 
+  val nestedListRelation = LocalRelation(
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TestRelations.scala]: will this relation be used in more tests? if not let's inline it in your test case.

Note that, the more common test relations we have, the harder to review test cases, as we need to search the definition of the used test relations.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: yea, ok.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TestRelations.scala (hunk 3 above)
Comment (maropu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TestRelations.scala]: ok, I'll remove this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: haha, yes. Currently, this method only checks exprs like `add(explode(v), literal(1))`, but does not check exprs like `expolode(explode(v))` handled in this pr.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 4]
```diff
@@ -2164,10 +2164,20 @@ class Analyzer(
     }
 
     private def hasNestedGenerator(expr: NamedExpression): Boolean = {
+      def hasUnsupportedInnerGenerator(g: Generator): Boolean = g match {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: what does "unsupported" mean here? Looking at the code it just matches `Generator`.

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: my bad... How about just `hasInnerGenerator `?

**Review by cloud-fan — APPROVED** [review]

**Comment by maropu** [comment]
Merged to master/branch-3.0. Thanks for the review, @cloud-fan! Could you check https://github.com/apache/spark/pull/27749, too?

**Comment by maropu** [comment]
I couldn't merge this into branch-2.4 cleanly, so I'll make a PR to backport it later. FYI: @dongjoon-hyun @HyukjinKwon

**Pull request closed without merging**